Add validation in policy files for missing codebases

rjernst opened the pull request:
Elasticsearch plugins can add a java security policy file to grant
additional permissions. These policy files can contain permission grants
for specific jar files, which are specified through system properties.
Unfortunately the java policy parser is lenient when a system property
is missing, meaning we can't know if there is a typo or grant for a no
longer relevant jar file.

This commit adds validation to the policy parsing by overriding the
system properties and tracking when a missing system property is used.

rjernst committed:
Add validation in policy files for missing codebases

Elasticsearch plugins can add a java security policy file to grant
additional permissions. These policy files can contain permission grants
for specific jar files, which are specified through system properties.
Unfortunately the java policy parser is lenient when a system property
is missing, meaning we can't know if there is a typo or grant for a no
longer relevant jar file.

This commit adds validation to the policy parsing by overriding the
system properties and tracking when a missing system property is used.

rjernst committed:
cleanup

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Core)

rjernst committed:
always add necessary codebases if they are needed

rjernst committed:
checkstyle

rjernst committed:
add test codebases file for plugins

rjernst committed:
account for policy file being in a jar

rjernst committed:
test codebases

rjernst committed:
checkstyle

rjernst committed:
remove debug

rjernst commented:
@elasticmachine run elasticsearch-ci/2

jaymode commented on test/framework/src/main/java/org/elasticsearch/bootstrap/BootstrapForTesting.java:
```diff
@@ -234,11 +239,30 @@ private static void addClassCodebase(Map<String, URL> codebases, String name, St
+
+            // if the codebases file is inside a jar, then we don't need to load it since the jar will
+            // have already been read from the classpath
+            if (policyFile.toString().contains(".jar") == false) {
```
I wonder for safety/strictness if we should change `.jar` to `.jar!`? Could also even include the path separator.

jaymode reviewed (approved):
LGTM

rjernst committed:
Isolate sql cli tests with dedicated testcase

The sql cli tests build on ESTestCase, but this triggers security
manager initialization for tests. However, this is unrealistic for the
sql cli both because it does not run with security manager, and the
cli only picks up the server jar that contains the policy and security
manager bootstrapping code unintentionally. Removing the depdnence on
the server code would be ideal, but until that can happen, this commit
starts the separation by adding a dedicated testcase for these tests
which does not use BootstrapForTesting.

rjernst committed:
Merge branch 'master' into plugin_permission_list6

rjernst committed:
Merge branch 'master' into plugin_permission_list6

rjernst committed:
refine jar FS check
